### ISSUE TITLE
Flattening monadic layers

### ISSUE THREAD

**Issue by int-index** [int-index/ether#13]
I believe that the performance could be improved greatly if there was a way to flatten monadic effects.

**Comment by int-index** [comment]
Ok, there's a working prototype for `ReaderT` now (and the benchmarks show that it's as fast as `mtl`).

**Comment by int-index** [comment]
[compound-types](http://hackage.haskell.org/package/compound-types) looks like a good candidate for Sum/Product types to use with flattening of implicitly tagged transformers.

**Comment by jberryman** [comment]
I'm just exploring this space, but there is also http://hackage.haskell.org/package/data-has which might be useful

**Comment by int-index** [comment]
Yes, I've seen it. What I don't like about the `data-has` package is that its `Has` class doesn't define a lens, it defines a getter and a setter. For types with non-constant access this is important, as a lens can traverse the structure just once, whereas with a getter/setter pair you'd have to do it twice (once to get a value, then to set a new one).
